File: python/Stack.py
from collections import deque
from collections import deque
import heapq
from collections import Counter
import bisect
import itertools


# 1. Implement Stack Using Arrays
class ArrayStack:

    def __init__(self):
        self.stack = []

    def push(self, x):
        # Add element to top of stack
        self.stack.append(x)

    def pop(self):
        # Remove and return top element
        if not self.is_empty():
            return self.stack.pop()
        return None

    def top(self):
        # Return top element without removing
        if not self.is_empty():
            return self.stack[-1]
        return None

    def is_empty(self):
        # Check if stack is empty
        return len(self.stack) == 0


# 2. Implement Stack Using Linked List
class ListNode:

    def __init__(self, val):
        self.val = val
        self.next = None


class LinkedListStack:

    def __init__(self):
        self.head = None

    def push(self, x):
        # Insert at head
        node = ListNode(x)
        node.next = self.head
        self.head = node

    def pop(self):
        # Remove head node
        if self.head:
            val = self.head.val
            self.head = self.head.next
            return val
        return None

    def top(self):
        # Return head value
        if self.head:
            return self.head.val
        return None

    def is_empty(self):
        return self.head is None


# 3. Design a Stack That Supports getMin() in O(1)
class MinStack:

    def __init__(self):
        self.stack = []
        self.min_stack = []

    def push(self, x):
        self.stack.append(x)
        # Push to min_stack if it's the new min
        if not self.min_stack or x <= self.min_stack[-1]:
            self.min_stack.append(x)

    def pop(self):
        if self.stack:
            val = self.stack.pop()
            if val == self.min_stack[-1]:
                self.min_stack.pop()
            return val
        return None

    def top(self):
        if self.stack:
            return self.stack[-1]
        return None

    def getMin(self):
        # Return current minimum
        if self.min_stack:
            return self.min_stack[-1]
        return None


# 4. Evaluate Reverse Polish Notation
def evalRPN(tokens):
    # Use stack to evaluate postfix expression
    stack = []
    for token in tokens:
        if token in "+-*/":
            b, a = stack.pop(), stack.pop()
            if token == '+': stack.append(a + b)
            elif token == '-': stack.append(a - b)
            elif token == '*': stack.append(a * b)
            elif token == '/': stack.append(int(a / b))
        else:
            stack.append(int(token))
    return stack[0]


# 5. Implement Queue Using Stacks
class MyQueue:

    def __init__(self):
        self.in_stack = []
        self.out_stack = []

    def push(self, x):
        self.in_stack.append(x)

    def pop(self):
        self.peek()
        return self.out_stack.pop()

    def peek(self):
        if not self.out_stack:
            while self.in_stack:
                self.out_stack.append(self.in_stack.pop())
        return self.out_stack[-1]

    def empty(self):
        return not self.in_stack and not self.out_stack


# 6. Valid Parentheses
def isValid(s):
    # Use stack to match parentheses
    stack = []
    mapping = {')': '(', '}': '{', ']': '['}
    for char in s:
        if char in mapping.values():
            stack.append(char)
        elif char in mapping:
            if not stack or stack.pop() != mapping[char]:
                return False
    return not stack


# 7. Generate Parentheses
def generateParenthesis(n):
    # Backtracking to generate all valid combinations
    res = []

    def backtrack(s, left, right):
        if len(s) == 2 * n:
            res.append(s)
            return
        if left < n:
            backtrack(s + '(', left + 1, right)
        if right < left:
            backtrack(s + ')', left, right + 1)

    backtrack('', 0, 0)
    return res


# 8. Longest Valid Parentheses
def longestValidParentheses(s):
    # Use stack to track indices of valid substrings
    stack = [-1]
    max_len = 0
    for i, c in enumerate(s):
        if c == '(':
            stack.append(i)
        else:
            stack.pop()
            if not stack:
                stack.append(i)
            else:
                max_len = max(max_len, i - stack[-1])
    return max_len


# 9. Remove Invalid Parentheses
def removeInvalidParentheses(s):
    # BFS to remove minimum invalid parentheses
    def is_valid(string):
        count = 0
        for c in string:
            if c == '(': count += 1
            if c == ')': count -= 1
            if count < 0: return False
        return count == 0

    visited = set([s])
    queue = deque([s])
    res = []
    found = False
    while queue:
        curr = queue.popleft()
        if is_valid(curr):
            res.append(curr)
            found = True
        if found: continue
        for i in range(len(curr)):
            if curr[i] not in '()': continue
            nxt = curr[:i] + curr[i + 1:]
            if nxt not in visited:
                visited.add(nxt)
                queue.append(nxt)
    return res


# 10. Score of Parentheses
def scoreOfParentheses(s):
    # Stack to compute score based on rules
    stack = [0]
    for c in s:
        if c == '(':
            stack.append(0)
        else:
            v = stack.pop()
            stack[-1] += max(2 * v, 1)
    return stack[0]


# 11. Next Greater Element I
def nextGreaterElement(nums1, nums2):
    # Monotonic stack to find next greater for each element
    stack, mapping = [], {}
    for num in nums2:
        while stack and num > stack[-1]:
            mapping[stack.pop()] = num
        stack.append(num)
    return [mapping.get(x, -1) for x in nums1]


# 12. Next Greater Element II (Circular Array)
def nextGreaterElements(nums):
    # Traverse array twice for circular effect
    n = len(nums)
    res = [-1] * n
    stack = []
    for i in range(2 * n):
        num = nums[i % n]
        while stack and nums[stack[-1]] < num:
            res[stack.pop()] = num
        if i < n:
            stack.append(i)
    return res


# 13. Daily Temperatures
def dailyTemperatures(T):
    # Monotonic stack to find next warmer day
    res = [0] * len(T)
    stack = []
    for i, temp in enumerate(T):
        while stack and T[stack[-1]] < temp:
            idx = stack.pop()
            res[idx] = i - idx
        stack.append(i)
    return res


# 14. Largest Rectangle in Histogram
def largestRectangleArea(heights):
    # Stack to maintain indices of increasing bars
    stack = []
    max_area = 0
    heights.append(0)
    for i, h in enumerate(heights):
        while stack and heights[stack[-1]] > h:
            height = heights[stack.pop()]
            width = i if not stack else i - stack[-1] - 1
            max_area = max(max_area, height * width)
        stack.append(i)
    heights.pop()
    return max_area


# 15. Maximal Rectangle
def maximalRectangle(matrix):
    # Use histogram approach for each row
    if not matrix: return 0
    n = len(matrix[0])
    heights = [0] * n
    max_area = 0
    for row in matrix:
        for i in range(n):
            heights[i] = heights[i] + 1 if row[i] == '1' else 0
        max_area = max(max_area, largestRectangleArea(heights))
    return max_area


# 16. Min Stack (Same as #3)
# See MinStack class above


# 17. Stock Span Problem
def calculateSpan(prices):
    # Stack to store indices of previous higher price
    n = len(prices)
    span = [1] * n
    stack = []
    for i in range(n):
        while stack and prices[i] >= prices[stack[-1]]:
            stack.pop()
        span[i] = i + 1 if not stack else i - stack[-1]
        stack.append(i)
    return span


# 18. Sliding Window Maximum
def maxSlidingWindow(nums, k):
    # Monotonic deque to keep max at front
    dq = deque()
    res = []
    for i, num in enumerate(nums):
        while dq and nums[dq[-1]] < num:
            dq.pop()
        dq.append(i)
        if dq[0] == i - k:
            dq.popleft()
        if i >= k - 1:
            res.append(nums[dq[0]])
    return res


# 19. Decode String
def decodeString(s):
    # Stack to decode nested patterns
    stack = []
    curr_num = 0
    curr_str = ''
    for c in s:
        if c.isdigit():
            curr_num = curr_num * 10 + int(c)
        elif c == '[':
            stack.append((curr_str, curr_num))
            curr_str, curr_num = '', 0
        elif c == ']':
            last_str, num = stack.pop()
            curr_str = last_str + num * curr_str
        else:
            curr_str += c
    return curr_str


# 20. Basic Calculator (I & II)
def calculate(s):
    # Evaluate expression with +, -, *, /
    stack = []
    num = 0
    sign = '+'
    s += '+'
    for c in s:
        if c.isdigit():
            num = num * 10 + int(c)
        elif c in '+-*/':
            if sign == '+':
                stack.append(num)
            elif sign == '-':
                stack.append(-num)
            elif sign == '*':
                stack.append(stack.pop() * num)
            elif sign == '/':
                stack.append(int(stack.pop() / num))
            num = 0
            sign = c
    return sum(stack)

    # 21. Largest Rectangle in Histogram
    # (Already implemented above as largestRectangleArea)
    # Explanation:
    # Use a stack to keep indices of increasing bar heights. When a lower bar is found,
    # pop from the stack and calculate area with the popped bar as the smallest bar.
    # This ensures all rectangles are considered.

    # 22. Maximal Rectangle in Binary Matrix
    # (Already implemented above as maximalRectangle)
    # Explanation:
    # Treat each row as the base of a histogram and use the largestRectangleArea function
    # for each row to find the maximal rectangle of 1's.


    # 23. Implement a Stack with Increment Operation
    # Design a stack that supports push, pop, and increment the bottom k elements by val.
class CustomStack:

    def __init__(self, maxSize):
        self.stack = []
        self.maxSize = maxSize

    def push(self, x):
        if len(self.stack) < self.maxSize:
            self.stack.append(x)

    def pop(self):
        if self.stack:
            return self.stack.pop()
        return -1

    def increment(self, k, val):
        for i in range(min(k, len(self.stack))):
            self.stack[i] += val


# 24. Simplify Path (Unix Style)
# Given a Unix-style path, simplify it (e.g., "/a/./b/../../c/" -> "/c")
def simplifyPath(path):
    stack = []
    for part in path.split('/'):
        if part == '' or part == '.':
            continue
        elif part == '..':
            if stack:
                stack.pop()
        else:
            stack.append(part)
    return '/' + '/'.join(stack)


# 25. Remove K Digits
# Remove k digits from the number to make it the smallest possible
def removeKdigits(num, k):
    stack = []
    for digit in num:
        while k and stack and stack[-1] > digit:
            stack.pop()
            k -= 1
        stack.append(digit)
    # Remove remaining digits from the end if k > 0
    final = stack[:-k] if k else stack
    # Remove leading zeros
    return ''.join(final).lstrip('0') or '0'


# 26. Trapping Rain Water Using Stack
# Given heights, compute how much water is trapped after raining
def trap(height):
    stack = []
    water = 0
    for i, h in enumerate(height):
        while stack and h > height[stack[-1]]:
            top = stack.pop()
            if not stack:
                break
            distance = i - stack[-1] - 1
            bounded_height = min(h, height[stack[-1]]) - height[top]
            water += distance * bounded_height
        stack.append(i)
    return water


# 27. Longest Valid Parentheses
# (Already implemented above as longestValidParentheses)
# Explanation:
# Use a stack to track indices of '(' and unmatched ')'. The difference between indices
# gives the length of valid substrings.


# 28. Design Hit Counter Using Stack
# Count hits in the past 5 minutes (300 seconds)
class HitCounter:

    def __init__(self):
        self.hits = []

    def hit(self, timestamp):
        self.hits.append(timestamp)

    def getHits(self, timestamp):
        # Remove hits older than 300 seconds
        while self.hits and self.hits[0] <= timestamp - 300:
            self.hits.pop(0)
        return len(self.hits)


# 29. Next Smaller Element
# For each element, find the next smaller element to its right
def nextSmallerElement(nums):
    stack = []
    res = [-1] * len(nums)
    for i in range(len(nums) - 1, -1, -1):
        while stack and stack[-1] >= nums[i]:
            stack.pop()
        if stack:
            res[i] = stack[-1]
        stack.append(nums[i])
    return res


# 30. Expression Add Operators
# Given num as a string and target, return all expressions by adding +, -, * that evaluate to target
def addOperators(num, target):
    res = []

    def backtrack(index, path, value, prev):
        if index == len(num):
            if value == target:
                res.append(path)
            return
        for i in range(index + 1, len(num) + 1):
            tmp = num[index:i]
            if len(tmp) > 1 and tmp[0] == '0':
                break
            n = int(tmp)
            if index == 0:
                backtrack(i, tmp, n, n)
            else:
                backtrack(i, path + '+' + tmp, value + n, n)
                backtrack(i, path + '-' + tmp, value - n, -n)
                backtrack(i, path + '*' + tmp, value - prev + prev * n,
                          prev * n)

    backtrack(0, '', 0, 0)
    return res

File: GO/Stacks/Stack.go
package stacks

import (
	"strconv"
	"strings"
)

// 1. Stack Using Arrays
type ArrayStack struct {
	data []int
}

func NewArrayStack() *ArrayStack {
	return &ArrayStack{data: []int{}}
}

func (s *ArrayStack) Push(x int) {
	s.data = append(s.data, x)
}

func (s *ArrayStack) Pop() int {
	if len(s.data) == 0 {
		panic("stack is empty")
	}
	val := s.data[len(s.data)-1]
	s.data = s.data[:len(s.data)-1]
	return val
}

func (s *ArrayStack) Top() int {
	if len(s.data) == 0 {
		panic("stack is empty")
	}
	return s.data[len(s.data)-1]
}

func (s *ArrayStack) IsEmpty() bool {
	return len(s.data) == 0
}

// 2. Stack Using Linked List
type StackListNode struct {
	val  int
	next *StackListNode
}

type LinkedListStack struct {
	head *StackListNode
}

func NewLinkedListStack() *LinkedListStack {
	return &LinkedListStack{nil}
}

func (s *LinkedListStack) Push(x int) {
	node := &StackListNode{val: x, next: s.head}
	s.head = node
}

func (s *LinkedListStack) Pop() int {
	if s.head == nil {
		panic("stack is empty")
	}
	val := s.head.val
	s.head = s.head.next
	return val
}

func (s *LinkedListStack) Top() int {
	if s.head == nil {
		panic("stack is empty")
	}
	return s.head.val
}

func (s *LinkedListStack) IsEmpty() bool {
	return s.head == nil
}

// 3. Stack with getMin() in O(1)
type MinStack struct {
	stack    []int
	minStack []int
}

func NewMinStack() *MinStack {
	return &MinStack{stack: []int{}, minStack: []int{}}
}

func (s *MinStack) Push(x int) {
	s.stack = append(s.stack, x)
	if len(s.minStack) == 0 || x <= s.minStack[len(s.minStack)-1] {
		s.minStack = append(s.minStack, x)
	}
}

func (s *MinStack) Pop() int {
	if len(s.stack) == 0 {
		panic("stack is empty")
	}
	val := s.stack[len(s.stack)-1]
	s.stack = s.stack[:len(s.stack)-1]
	if val == s.minStack[len(s.minStack)-1] {
		s.minStack = s.minStack[:len(s.minStack)-1]
	}
	return val
}

func (s *MinStack) Top() int {
	if len(s.stack) == 0 {
		panic("stack is empty")
	}
	return s.stack[len(s.stack)-1]
}

func (s *MinStack) GetMin() int {
	if len(s.minStack) == 0 {
		panic("stack is empty")
	}
	return s.minStack[len(s.minStack)-1]
}

// 4. Evaluate Reverse Polish Notation
func EvalRPN(tokens []string) int {
	stack := []int{}
	for _, token := range tokens {
		switch token {
		case "+", "-", "*", "/":
			b := stack[len(stack)-1]
			a := stack[len(stack)-2]
			stack = stack[:len(stack)-2]
			switch token {
			case "+":
				stack = append(stack, a+b)
			case "-":
				stack = append(stack, a-b)
			case "*":
				stack = append(stack, a*b)
			case "/":
				stack = append(stack, a/b)
			}
		default:
			num, _ := strconv.Atoi(token)
			stack = append(stack, num)
		}
	}
	return stack[0]
}

// 5. Queue Using Stacks
type QueueWithStacks struct {
	inStack  []int
	outStack []int
}

func NewQueueWithStacks() *QueueWithStacks {
	return &QueueWithStacks{}
}

func (q *QueueWithStacks) Push(x int) {
	q.inStack = append(q.inStack, x)
}

func (q *QueueWithStacks) Pop() int {
	if len(q.outStack) == 0 {
		for len(q.inStack) > 0 {
			n := len(q.inStack)
			q.outStack = append(q.outStack, q.inStack[n-1])
			q.inStack = q.inStack[:n-1]
		}
	}
	if len(q.outStack) == 0 {
		panic("queue is empty")
	}
	val := q.outStack[len(q.outStack)-1]
	q.outStack = q.outStack[:len(q.outStack)-1]
	return val
}

func (q *QueueWithStacks) Peek() int {
	if len(q.outStack) == 0 {
		for len(q.inStack) > 0 {
			n := len(q.inStack)
			q.outStack = append(q.outStack, q.inStack[n-1])
			q.inStack = q.inStack[:n-1]
		}
	}
	if len(q.outStack) == 0 {
		panic("queue is empty")
	}
	return q.outStack[len(q.outStack)-1]
}

func (q *QueueWithStacks) Empty() bool {
	return len(q.inStack) == 0 && len(q.outStack) == 0
}

// 6. Valid Parentheses
func IsValidParentheses(s string) bool {
	stack := []rune{}
	mapping := map[rune]rune{')': '(', '}': '{', ']': '['}
	for _, ch := range s {
		if ch == '(' || ch == '{' || ch == '[' {
			stack = append(stack, ch)
		} else {
			if len(stack) == 0 || stack[len(stack)-1] != mapping[ch] {
				return false
			}
			stack = stack[:len(stack)-1]
		}
	}
	return len(stack) == 0
}

// 7. Generate Parentheses
func GenerateParentheses(n int) []string {
	var res []string
	var backtrack func(cur string, open, close int)
	backtrack = func(cur string, open, close int) {
		if len(cur) == 2*n {
			res = append(res, cur)
			return
		}
		if open < n {
			backtrack(cur+"(", open+1, close)
		}
		if close < open {
			backtrack(cur+")", open, close+1)
		}
	}
	backtrack("", 0, 0)
	return res
}

// 8. Longest Valid Parentheses
func LongestValidParentheses(s string) int {
	stack := []int{-1}
	maxLen := 0
	for i, ch := range s {
		if ch == '(' {
			stack = append(stack, i)
		} else {
			stack = stack[:len(stack)-1]
			if len(stack) == 0 {
				stack = append(stack, i)
			} else {
				if i-stack[len(stack)-1] > maxLen {
					maxLen = i - stack[len(stack)-1]
				}
			}
		}
	}
	return maxLen
}

// 9. Remove Invalid Parentheses
func RemoveInvalidParentheses(s string) []string {
	var res []string
	var dfs func(string, int, int, []rune)
	dfs = func(s string, last_i, last_j int, par []rune) {
		count := 0
		for i := last_i; i < len(s); i++ {
			if rune(s[i]) == par[0] {
				count++
			}
			if rune(s[i]) == par[1] {
				count--
			}
			if count >= 0 {
				continue
			}
			for j := last_j; j <= i; j++ {
				if rune(s[j]) == par[1] && (j == last_j || s[j-1] != s[j]) {
					dfs(s[:j]+s[j+1:], i, j, par)
				}
			}
			return
		}
		reversed := reverseString(s)
		if par[0] == '(' {
			dfs(reversed, 0, 0, []rune{')', '('})
		} else {
			res = append(res, reversed)
		}
	}
	dfs(s, 0, 0, []rune{'(', ')'})
	return uniqueStrings(res)
}

func reverseString(s string) string {
	runes := []rune(s)
	for i, j := 0, len(runes)-1; i < j; i, j = i+1, j-1 {
		runes[i], runes[j] = runes[j], runes[i]
	}
	return string(runes)
}

func uniqueStrings(arr []string) []string {
	m := make(map[string]struct{})
	var res []string
	for _, v := range arr {
		if _, ok := m[v]; !ok {
			m[v] = struct{}{}
			res = append(res, v)
		}
	}
	return res
}

// 10. Score of Parentheses
func ScoreOfParentheses(s string) int {
	stack := []int{0}
	for _, ch := range s {
		if ch == '(' {
			stack = append(stack, 0)
		} else {
			v := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			stack[len(stack)-1] += max(2*v, 1)
		}
	}
	return stack[0]
}

// 11. Next Greater Element I
func NextGreaterElement(nums1 []int, nums2 []int) []int {
	res := make([]int, len(nums1))
	m := make(map[int]int)
	stack := []int{}
	for i := len(nums2) - 1; i >= 0; i-- {
		for len(stack) > 0 && stack[len(stack)-1] <= nums2[i] {
			stack = stack[:len(stack)-1]
		}
		if len(stack) == 0 {
			m[nums2[i]] = -1
		} else {
			m[nums2[i]] = stack[len(stack)-1]
		}
		stack = append(stack, nums2[i])
	}
	for i, v := range nums1 {
		res[i] = m[v]
	}
	return res
}

// 12. Next Greater Element II (Circular Array)
func NextGreaterElementsCircular(nums []int) []int {
	n := len(nums)
	res := make([]int, n)
	for i := range res {
		res[i] = -1
	}
	stack := []int{}
	for i := 0; i < 2*n; i++ {
		num := nums[i%n]
		for len(stack) > 0 && nums[stack[len(stack)-1]] < num {
			res[stack[len(stack)-1]] = num
			stack = stack[:len(stack)-1]
		}
		if i < n {
			stack = append(stack, i)
		}
	}
	return res
}

// 13. Daily Temperatures
func DailyTemperatures(temperatures []int) []int {
	n := len(temperatures)
	res := make([]int, n)
	stack := []int{}
	for i := 0; i < n; i++ {
		for len(stack) > 0 && temperatures[i] > temperatures[stack[len(stack)-1]] {
			idx := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			res[idx] = i - idx
		}
		stack = append(stack, i)
	}
	return res
}

// 14. Largest Rectangle in Histogram
func LargestRectangleInHistogram(heights []int) int {
	stack := []int{}
	maxArea := 0
	heights = append(heights, 0)
	for i, h := range heights {
		for len(stack) > 0 && h < heights[stack[len(stack)-1]] {
			height := heights[stack[len(stack)-1]]
			stack = stack[:len(stack)-1]
			width := i
			if len(stack) > 0 {
				width = i - stack[len(stack)-1] - 1
			}
			area := height * width
			if area > maxArea {
				maxArea = area
			}
		}
		stack = append(stack, i)
	}
	return maxArea
}

// 15. Maximal Rectangle
func MaximalRectangle(matrix [][]byte) int {
	if len(matrix) == 0 {
		return 0
	}
	n := len(matrix[0])
	heights := make([]int, n)
	maxArea := 0
	for _, row := range matrix {
		for i := 0; i < n; i++ {
			if row[i] == '1' {
				heights[i]++
			} else {
				heights[i] = 0
			}
		}
		area := LargestRectangleInHistogram(heights)
		if area > maxArea {
			maxArea = area
		}
	}
	return maxArea
}

// 16. Min Stack (already implemented as MinStack above)

// 17. Stock Span Problem
func StockSpan(prices []int) []int {
	n := len(prices)
	res := make([]int, n)
	stack := []int{}
	for i := 0; i < n; i++ {
		for len(stack) > 0 && prices[i] >= prices[stack[len(stack)-1]] {
			stack = stack[:len(stack)-1]
		}
		if len(stack) == 0 {
			res[i] = i + 1
		} else {
			res[i] = i - stack[len(stack)-1]
		}
		stack = append(stack, i)
	}
	return res
}

// 18. Sliding Window Maximum
func SlidingWindowMaximum(nums []int, k int) []int {
	var res []int
	deque := []int{}
	for i := 0; i < len(nums); i++ {
		if len(deque) > 0 && deque[0] <= i-k {
			deque = deque[1:]
		}
		for len(deque) > 0 && nums[i] >= nums[deque[len(deque)-1]] {
			deque = deque[:len(deque)-1]
		}
		deque = append(deque, i)
		if i >= k-1 {
			res = append(res, nums[deque[0]])
		}
	}
	return res
}

// 19. Decode String
func DecodeString(s string) string {
	countStack := []int{}
	strStack := []string{}
	curStr := ""
	k := 0
	for _, ch := range s {
		if ch >= '0' && ch <= '9' {
			k = k*10 + int(ch-'0')
		} else if ch == '[' {
			countStack = append(countStack, k)
			strStack = append(strStack, curStr)
			curStr = ""
			k = 0
		} else if ch == ']' {
			n := countStack[len(countStack)-1]
			countStack = countStack[:len(countStack)-1]
			prevStr := strStack[len(strStack)-1]
			strStack = strStack[:len(strStack)-1]
			curStr = prevStr + strings.Repeat(curStr, n)
		} else {
			curStr += string(ch)
		}
	}
	return curStr
}

// 20. Basic Calculator I
func BasicCalculator(s string) int {
	stack := []int{}
	num := 0
	sign := 1
	res := 0
	for i, ch := range s {
		if ch >= '0' && ch <= '9' {
			num = num*10 + int(ch-'0')
		}
		if (ch < '0' && ch != ' ') || i == len(s)-1 {
			res += sign * num
			num = 0
			if ch == '+' {
				sign = 1
			} else if ch == '-' {
				sign = -1
			} else if ch == '(' {
				stack = append(stack, res)
				stack = append(stack, sign)
				res = 0
				sign = 1
			} else if ch == ')' {
				sign = stack[len(stack)-1]
				stack = stack[:len(stack)-1]
				res = stack[len(stack)-1] + sign*res
				stack = stack[:len(stack)-1]
			}
		}
	}
	return res
}

// 21. Largest Rectangle in Histogram (already implemented as LargestRectangleInHistogram)

// 22. Maximal Rectangle in Binary Matrix (already implemented as MaximalRectangle)

// 23. Stack with Increment Operation
type CustomStack struct {
	stack []int
	max   int
}

func NewCustomStack(maxSize int) *CustomStack {
	return &CustomStack{stack: []int{}, max: maxSize}
}

func (s *CustomStack) Push(x int) {
	if len(s.stack) < s.max {
		s.stack = append(s.stack, x)
	}
}

func (s *CustomStack) Pop() int {
	if len(s.stack) == 0 {
		return -1
	}
	val := s.stack[len(s.stack)-1]
	s.stack = s.stack[:len(s.stack)-1]
	return val
}

func (s *CustomStack) Increment(k int, val int) {
	n := min(k, len(s.stack))
	for i := 0; i < n; i++ {
		s.stack[i] += val
	}
}

// 24. Simplify Path (Unix Style)
func SimplifyPath(path string) string {
	parts := strings.Split(path, "/")
	stack := []string{}
	for _, part := range parts {
		if part == "" || part == "." {
			continue
		}
		if part == ".." {
			if len(stack) > 0 {
				stack = stack[:len(stack)-1]
			}
		} else {
			stack = append(stack, part)
		}
	}
	return "/" + strings.Join(stack, "/")
}

// 25. Remove K Digits
func RemoveKDigits(num string, k int) string {
	stack := []byte{}
	for i := 0; i < len(num); i++ {
		for k > 0 && len(stack) > 0 && stack[len(stack)-1] > num[i] {
			stack = stack[:len(stack)-1]
			k--
		}
		stack = append(stack, num[i])
	}
	stack = stack[:len(stack)-k]
	res := strings.TrimLeft(string(stack), "0")
	if res == "" {
		return "0"
	}
	return res
}

/*
26. Trapping Rain Water Using Stack
*/
func TrapRainWater(height []int) int {
	stack := []int{}
	water := 0
	for i, h := range height {
		for len(stack) > 0 && h > height[stack[len(stack)-1]] {
			top := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			if len(stack) == 0 {
				break
			}
			distance := i - stack[len(stack)-1] - 1
			boundedHeight := min(h, height[stack[len(stack)-1]]) - height[top]
			water += distance * boundedHeight
		}
		stack = append(stack, i)
	}
	return water
}

/*
27. Longest Valid Parentheses (Stack-based)
*/
func LongestValidParenthesesStack(s string) int {
	stack := []int{}
	maxLen := 0
	lastInvalid := -1
	for i, ch := range s {
		if ch == '(' {
			stack = append(stack, i)
		} else {
			if len(stack) == 0 {
				lastInvalid = i
			} else {
				stack = stack[:len(stack)-1]
				if len(stack) == 0 {
					maxLen = max(maxLen, i-lastInvalid)
				} else {
					maxLen = max(maxLen, i-stack[len(stack)-1])
				}
			}
		}
	}
	return maxLen
}

/*
28. Design Hit Counter Using Stack
*/
type HitCounter struct {
	times []int
}

func NewHitCounter() *HitCounter {
	return &HitCounter{times: []int{}}
}

func (h *HitCounter) Hit(timestamp int) {
	h.times = append(h.times, timestamp)
}

func (h *HitCounter) GetHits(timestamp int) int {
	for len(h.times) > 0 && h.times[0] <= timestamp-300 {
		h.times = h.times[1:]
	}
	return len(h.times)
}

/*
29. Next Smaller Element
*/
func NextSmallerElement(nums []int) []int {
	n := len(nums)
	res := make([]int, n)
	stack := []int{}
	for i := n - 1; i >= 0; i-- {
		for len(stack) > 0 && stack[len(stack)-1] >= nums[i] {
			stack = stack[:len(stack)-1]
		}
		if len(stack) == 0 {
			res[i] = -1
		} else {
			res[i] = stack[len(stack)-1]
		}
		stack = append(stack, nums[i])
	}
	return res
}

/*
30. Expression Add Operators
*/
func AddOperators(num string, target int) []string {
	var res []string
	var dfs func(path string, pos int, eval int, multed int)
	dfs = func(path string, pos int, eval int, multed int) {
		if pos == len(num) {
			if eval == target {
				res = append(res, path)
			}
			return
		}
		for i := pos; i < len(num); i++ {
			if i != pos && num[pos] == '0' {
				break
			}
			curStr := num[pos : i+1]
			cur, _ := strconv.Atoi(curStr)
			if pos == 0 {
				dfs(curStr, i+1, cur, cur)
			} else {
				dfs(path+"+"+curStr, i+1, eval+cur, cur)
				dfs(path+"-"+curStr, i+1, eval-cur, -cur)
				dfs(path+"*"+curStr, i+1, eval-multed+multed*cur, multed*cur)
			}
		}
	}
	dfs("", 0, 0, 0)
	return res
}
